avm/src/vm/memory.rs

```diff
@@ -136,67 +144,120 @@ impl HandlerMemory {
     }
   }
 
-  /// Takes a given address and looks up the `mems` indexes relevant to it.
-  fn addr_to_idxs(self: &HandlerMemory, addr: i64) -> (usize, usize) {
+  pub fn drop_parent(self: &mut HandlerMemory) {
+    if self.parent.is_some() {
+      let arc = self.parent.take();
+      drop(arc);
+    }
+  }
+
+  /// Returns true if the `mem` indexes are defined within this HandlerMemory
+  fn are_idxs_defined(self: &HandlerMemory, a: usize, b: usize) -> bool {
+    if a == std::usize::MAX {
+      return true;
+    } else if self.mems.len() > a && b == std::usize::MAX {
+      return true;
+    } else if self.mems.len() > a && self.mems[a].len() > b {
+      return true;
+    }
+    return false;
+  }
+
+  /// Takes a given address and finds out the `mems` indexes and if they are defined
+  fn addr_to_idxs_opt(self: &HandlerMemory, addr: i64) -> Option<(usize, usize)> {
     return if addr >= 0 {
-     self.addr.0[addr as usize]
+      *self.addr.0.get(addr as usize).unwrap_or(&None)
     } else if addr <= CLOSURE_ARG_MEM_END {
-      self.addr.1[(addr - CLOSURE_ARG_MEM_START) as usize]
+      *self.addr.1.get((addr - CLOSURE_ARG_MEM_START) as usize).unwrap_or(&None)
     } else {
-      (0, ((-1 * addr - 1) / 8) as usize)
-    }
+      Some((0, ((-1 * addr - 1) / 8) as usize))
+    };
+  }
+
+  /// Takes a given address and looks up the fractal location and
+  /// `mems` indexes relevant to it. It also returns a boolean that is
+  /// true if the address points to, or is in, the parent
+  fn addr_to_idxs(self: &HandlerMemory, addr: i64) -> ((usize, usize), bool) {
+    let idxs = self.addr_to_idxs_opt(addr);
+    return if idxs.is_none() {
+      // fail if no parent
+      if self.parent.is_none() {
+        panic!("Memory address referenced in parent, but no parent pointer defined");
+      }
+      // fail if parent is present, but idx is not defined
+      let hm = self.parent.as_ref().unwrap();
+      let idxs = hm.addr_to_idxs_opt(addr);
+      if idxs.is_none() {
+        panic!("Memory address referenced in parent, but no indices correspond to the virtual address");
+      }
+      (idxs.unwrap(), true)
```

Comment:
This'll fail if you're multiple levels deep, eg double-nested `map`. You should just call `addr_to_idxs` on the parent hm in that case and let it handle it recursively, rather than having special case logic for the parent.